alf/environments/parallel_torch_environment.py

```diff
@@ -0,0 +1,188 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Runs multiple environments in parallel processes and steps them in batch.
+
+Adapted from TF-Agents Environment API as seen in:
+    https://github.com/tensorflow/agents/blob/master/tf_agents/environments/parallel_py_environment.py
+"""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+from absl import logging
+
+import gin
+import torch
+
+from alf.environments import torch_environment
+from alf.environments.process_environment import ProcessEnvironment
+import alf.nest as nest
+
+
+@gin.configurable
+class ParallelTorchEnvironment(torch_environment.TorchEnvironment):
+    """Batch together environments and simulate them in external processes.
+
+  The environments are created in external processes by calling the provided
+  callables. This can be an environment class, or a function creating the
+  environment and potentially wrapping it. The returned environment should not
+  access global variables.
+  """
+
+    def __init__(self,
+                 env_constructors,
+                 start_serially=True,
+                 blocking=False,
+                 flatten=False):
+        """Batch together environments and simulate them in external processes.
+
+    The environments can be different but must use the same action and
+    observation specs.
+
+    Args:
+      env_constructors: List of callables that create environments.
+      start_serially: Whether to start environments serially or in parallel.
+      blocking: Whether to step environments one after another.
+      flatten: Boolean, whether to use flatten action and time_steps during
+        communication to reduce overhead.
+
+    Raises:
+      ValueError: If the action or observation specs don't match.
+    """
+        super(ParallelTorchEnvironment, self).__init__()
+        self._envs = []
+        self._env_ids = []
+        for env_id, ctor in enumerate(env_constructors):
+            env = ProcessEnvironment(ctor, flatten=flatten)
+            env_id = torch.tensor(env_id, dtype=torch.int32)
+            env._env_id = env_id
+            self._envs.append(env)
+            self._env_ids.append(env_id)
+        self._num_envs = len(env_constructors)
+        self._blocking = blocking
+        self._start_serially = start_serially
+        self.start()
+        self._action_spec = self._envs[0].action_spec()
```

Comment:
just wondering, isn't it strange that the specs are not batched tensor specs, but instead the specs of the individual env?